### ISSUE TITLE
question of the 2nd term of DDC

### ISSUE THREAD

**Issue by iTomxy** [DanielTrosten/mvc#1]
I think there maybe a mistake in [DDC2](https://github.com/DanielTrosten/mvc/blob/main/src/lib/kernel.py#L86), although [a flipped version](https://github.com/DanielTrosten/mvc/blob/main/src/lib/kernel.py#L95) is provided.
To be clear, I mean the term that encourages the orthogonality between cluster assignments. Given an assignment matrix `A` with shape `[n, c]`, with `n` the number of samples and `c` the number of clusters, the loss should actually be calculated as:
```python
loss = 2 / (n * (n - 1)) * triu(matmul(transpose(A), A))
```
because each sample shall exclusively belong to exactly ONE cluster, but different samples can belong to a same cluster.
And this term is consistent in [1] and [2]. I guess you might have been misled by the shape statement of `A` in [1], which is definitely a mistake in the original paper because it disagree with the following formula.

---

1. Deep divergence-based approach to clustering
2. End-to-end adversarial-attention network for multi-modal clustering

**Comment by DanielTrosten** [comment]
I can assure you that the code, our paper on MVC [1], and the DDC paper [2] is correct. 

Choosing between `A @ A.T` and `A.T @ A` really comes down to whether or not you want the additional regularizing effect of the first option, or if it is sufficient to only enforce exclusivity between clusters. In our paper on MVC [1], we found the added regularization to be beneficial for the clustering performance.

The DDC paper [2] includes a small discussion on this topic:
>This also has the effect of a regularization term if the number of clusters is smaller than the number of input data points. Indeed, not all data points in the restricted space can be orthogonal to each other, forcing data points to repel each other, thereby acting against the terms that try to improve similarity. This term also encourages a balanced distribution of data points in the different classes, which makes our loss ideal for problems with balanced classes.

As you point out, there are works that use the `A.T @ A`-formulation, such as EAMC [3] and our paper on Recurrent DDC [4].

---

1. D. J. Trosten, S. Løkse, R. Jenssen, and M. Kampffmeyer, “Reconsidering Representation Alignment for Multi-view Clustering,” CVPR, 2021.
3. M. Kampffmeyer, S. Løkse, F. M. Bianchi, L. Livi, A.-B. Salberg, and R. Jenssen, “Deep Divergence-based Approach to Clustering,” Neural Networks, vol. 113, pp. 91–101, 2019.
5. R. Zhou and Y.-D. Shen, “End-to-End Adversarial-Attention Network for Multi-Modal Clustering,” CVPR, 2020.
7. D. J. Trosten, A. S. Strauman, M. Kampffmeyer, and R. Jenssen, “Recurrent Deep Divergence-Based Clustering for Simultaneous Feature Learning and Clustering of Variable Length Time Series,” ICASSP, 2019.
